### PR TITLE
Bug fix: Correctly return value for `foo = await bar` in console

### PR DESCRIPTION
**Pull request by haltman-at**
Addresses #3349.

This PR fixes the issue where expressions of the form `foo = await bar`, when entered in the Truffle Console, would perform the assignment but then not return the resulting value.  Note that this PR does **not** fix the more general issues with the use of `await` in the Truffle Console, which is issue #1546.  Fixing the broader problems are probably going to have to wait until we have removed support for both Node 12 and Node 14.

The fundamental problem, and the reason for #1546, is that all our `await`-handling code in the console is a bunch of hacks.  The ultimate solution to all of this, once we don't need to support Node 12 and Node 14 anymore, will be to remove all these hacks and do things the straightforward way that Node 16 will allow us to do it.  However, in the meantime, I still wanted to fix the particular issue of #3349.  To accomplish that, this PR uses... more hacks.  But, it's not worse than the already existing hacks surrounding it, so I think it's fine.

The reason that we get `undefined` in this case is because we have special handling for this particular case, and that special handling involves creating a little assignment script and then passing it to a VM to run; we do get the return value from our script, but the problem is that we wrote the script in a way such that it always returns `undefined`.  So, I've modified it to (conditionally) return the assigned value instead.

Specifically, if we are doing a bare assignment (so no `let` or `const`, those shouldn't return values), I modify the assignment script by appending the variable name to the end of it, so that the value is returned.  This solves the problem!  Or the narrow problem of #3349, anyway.  The broad problem of #1546 will have to wait...

(Although, you can do things beyond *just* `foo = await bar` with it... e.g., `x = (await web3.eth.net.getId()) + 1` will work fine!)

I didn't include any tests in this PR because um I'm not sure how to test this programmatically.  But potentially tests could be added if people think it's necessary (and if anyone can give me an idea for how, or point me to where we already do this sort of thing...)

### CODE REVIEW

**Review by cds-amal — COMMENTED** [review]
Nice one @haltman-at! I'll be able to review this tomorrow. This can be tested by utilizing the same setup for [console/develop REPL tests.](https://github.com/trufflesuite/truffle/blob/develop/packages/truffle/test/scenarios/commands/develop.js#L64-L78)

```javascript
    assert.equal(
      output.indexOf('undefined'),
      -1,
      "should not print `undefined` on naked assignments"
    );

```

**Comment by haltman-at** [comment]
Oh, I see!  I only looked at console, not develop, so I didn't see that.  Cool, I can go back and add a test then!  (Of course I won't be abusing `indexOf` like that, sheesh, we have better ways these days. :P )

**Review by cds-amal — CHANGES_REQUESTED** [review]
Nice one Harry. It's a good approach given the constraints we have. Of course, it would be nice to get a few tests in.

File: packages/core/lib/console.js [hunk 1]
```diff
@@ -356,6 +356,18 @@ class Console extends EventEmitter {
       assignment = assign
         ? `${assign.trim()} global.${RESULT}; void delete global.${RESULT};`
         : null;
+
+      //finally, if the assignment did not use var, const, or let, make sure to
+      //return the result afterward
+      if (assign) {
+        const bareAssignmentMatch = assign.match(
+          /^\s*([a-zA-Z_$][0-9a-zA-Z_$]*)\s*=\s*/
```
[the remark this answers is not in the document]
Comment (haltman-at) [reply packages/core/lib/console.js:364]: There's an `if (assign)` around the whole thing.  This code only kicks in if you're already in the special case for handling assignment expressions.  So I think the `===` concern is not a problem; the code simply won't be invoked in that case.

You're right that destructuring assignments won't work properly; but maybe that's fine for now?  Alternatively I could change it instead to check instead that things don't begin with `let`, `var`, or `const` (plus appropriate whitespace), if you think that's worth it; I'm a little wary of that, but likely it's fine?

[the remark this answers is not in the document]
Comment (cds-amal) [reply packages/core/lib/console.js:364]: I verified destructuring fails in production, so I think it's fine. These issues will go away when we have the proper tools to handle await correctly. I agree about avoiding `let/var/const` regex checks. So, I'm going to click approve on this one!

**Comment by haltman-at** [comment]
OK, added the test!

**Pull request merged**